### DISCUSSION TITLE
Include et al. to output of \cite command with \nonumcitations

### DISCUSSION THREAD

**Discussion by pabloabur** [olsak/OpTeX#127] [Q&A] [answered]
Hi,

I've been playing with \_citeA and \_printsavedcites, hoping to insert "et al." between author name and year (e.g. Last et al., 2023 instead of default Last, 2023), but with no success. Do you have any similar examples or suggestions about this? I tried looking at the code and documentation, but I guess my TeX skills are not good enough!

**Comment by olsak** [comment]
Please, give a minimal example. Then I can get this code and start playing with it. And I will better understand the context, what do you want to achieve.

**Reply by pabloabur** [reply]
I'm coming from ConTeXt, and there I could specify that if there are three authors or more, the output of **\cite** would be the first author only, followed by _et al., year_. So in the MWE below, I get

Cite 3 authors: [Lastone, 2023]
Cite 4 authors: [Lastfour, 2023]

But what I wanted was

Cite 3 authors: (Lastone, Lasttwo, and Lasthree, 2023)
Cite 4 authors: (Lastfour et al., 2023)

And I also wanted a way to print only the year within parenthesis e.g. Lastfour et al. (2023). I could perfectly do this by using **\ecite** or **\bib[label]={mark}**, but I was wondering whether I could use \**_namecount** and **\bibmark** to define an alternative **\cite** command. It could avoid typos and inconsistent citations across the text, but is it worth the effort?

Thank you for being so helpful

MWE
test.tex:
```
\fontfam[LMfonts]
\nonumcitations

Cite 3 authors: \cite[short]\par
Cite 4 authors: \cite[long]

\bigskip
\usebib/s (simple) biblist
\bye
```

biblist.bib:
```
@ARTICLE{short,
    title="Short title",
    author="Lastone, One and Lasttwo, Two and Lastthree, three",
    year="2023"
    }

@ARTICLE{long,
    title="Long title",
    author="Lastfour, Four and Lastfive, Five and Lastsix, Six and Lastseven, Seven",
    year="2023"
    }
```

**Comment by olsak** [comment]
Now, I understand the problem. You means `\cite` outputs when `\nonumcitations` is set. On the other hand, the "et al." is printed as expected in bib-reference list, we have the aumax option for this case. 

I'll look at this problem soon.

**Comment by olsak** [comment, marked as answer]
I tried to solve it by the commit https://github.com/olsak/OpTeX/commit/2e5741bff4015d98ddec5891aa7f8a6e62e354c0

**Reply by pabloabur** [reply]
Hi, thanks for taking the time to have a look at this. Unfortunately, the result is different from what I expected.  In the previous example, it prints:

Cite 3 authors: Lastone, 2023
Cite 4 authors: Lastfour, 2023

I manually installed the latest release and added the code below (as in OpTeX tricks) to test.tex:

```
\bibtexhook{
   \def\_createbibmark #1;#2,#3;#4,#5;#6,#7;#8,#9\_fin{%
      \ifx ^#8^%
        \ifx ^#6^%
           \ifx ^#4^#2%
           \else #2, and #4\fi
        \else #2, #4, and #6\fi
      \else #2, et al.\fi
      \space (#1)% 
   }
}
\_def\cite[#1]{{\_citeA#1,,,\_printsavedcites}} % no square brackets printed around
```

The logs generated were:

```
This is LuaTeX, Version 1.15.0 (TeX Live 2022/Homebrew) 
 restricted system commands enabled.
(./test.tex This is OpTeX (Olsak's Plain TeX), version <1.11 Mar 2023>
(./test.ref)
(/usr/local/Cellar/texlive/20220321_4/share/texmf-dist/tex/optex/base/f-lmfonts
.opm
FONT: [Latin Modern] -- \LMfonts (TeX Gyre fonts based on Coputer Modern)
 mods:{\roman \sans \quotset \ttset \ttprop \ttlight \ttcond \upital \dunhill s
ubmods:\caps \slant \nbold \bolder } vars:{\rm \bf \it \bi \tt } math:{LM}
MATH-FONT: "[latinmodern-math]" -- unicode math prepared.

(/usr/local/Cellar/texlive/20220321_4/share/texmf-dist/tex/optex/base/unimath-c
odes.opm
(/usr/local/Cellar/texlive/20220321_4/share/texmf-dist/tex/optex/base/mathclass
.opm)
(/usr/local/Cellar/texlive/20220321_4/share/texmf-dist/tex/optex/base/unimath-t
able.opm)))
(/usr/local/Cellar/texlive/20220321_4/share/texmf-dist/tex/optex/base/usebib.op
m
(/usr/local/Cellar/texlive/20220321_4/share/texmf-dist/tex/generic/librarian/li
brarian.tex))
(/usr/local/Cellar/texlive/20220321_4/share/texmf-dist/tex/optex/base/bib-iso69
0.opm) (./biblist.bib)
WARNING l.35: Missing field "journal" in [long].
WARNING l.35: Missing field "pages" in [long].
WARNING l.35: Missing field "journal" in [short].
WARNING l.35: Missing field "pages" in [short].
[1{/usr/local/Cellar/texlive/20220321_4/share/texmf-dist/fonts/map/pdftex/updma
p/pdftex.map}])</usr/local/Cellar/texlive/20220321_4/share/texmf-dist/fonts/ope
ntype/public/lm/lmromancaps10-regular.otf></usr/local/Cellar/texlive/20220321_4
/share/texmf-dist/fonts/opentype/public/lm/lmroman10-regular.otf>
Output written on test.pdf (1 page, 9167 bytes).
Transcript written on test.log.
```

**Reply by olsak** [reply]
Hi,

I tested this using your `biblist.bib` and using the file:

```
\fontfam[LMfonts]
\nonumcitations

\bibtexhook{
   \def\_createbibmark #1;#2,#3;#4,#5;#6,#7;#8,#9\_fin{%
      \ifx ^#8^%
        \ifx ^#6^%
           \ifx ^#4^#2%
           \else #2, and #4\fi
        \else #2, #4, and #6\fi
      \else #2, et al.\fi
      \space (#1)% 
   }
}
\_def\cite[#1]{{\_citeA#1,,,\_printsavedcites}} % no brackets printed around

Cite 3 authors: \cite[short]\par
Cite 4 authors: \cite[long]

\bigskip
\usebib/s (iso690) biblist
\bye
```

and the result is (after two OpTeX runs):

> Cite 3 authors: Lastone, Lasttwo, and Lastthree (2023)
>
> Cite 4 authors: Lastfour, et al. (2023)

Sorry, I am unable to find out where is the difference between your computer and mine. Maybe, your files
```
/usr/local/Cellar/texlive/20220321_4/share/texmf-dist/tex/optex/base/bib-iso690.opm
/usr/local/Cellar/texlive/20220321_4/share/texmf-dist/tex/optex/base/bib-smple.opm
```
are old ones, no version 2023-03-24. You didn't show our log file, only terminal output. The log file shows the version dates too:
```
...
(/home/olsak/texmf/tex/optex/base/usebib.opm
@:[usebib] Reading bib databases <2022-02-04>
(/usr/local/texlive/2022/texmf-dist/tex/generic/librarian/librarian.tex
\lb@entrycount=\count301
\lb@namecount=\count302
\lb@tempcount=\count303
\lb@sortnumber=\count304
\NameCount=\count305
\lb@read=\read2
(./pok.lbr))) (/home/olsak/texmf/tex/optex/base/bib-iso690.opm
@:[bib-iso690] BIB style (iso690) <2023-03-24>
) (./biblist.bib)
WARNING l.21: Missing field "journal" in [long].
...
```

**Reply by pabloabur** [reply]
You are right. My log file shows that I am still using the old bib version:

```
This is LuaTeX, Version 1.15.0 (TeX Live 2022/Homebrew)  (format=optex 2023.3.25)  26 MAR 2023 10:23
 restricted system commands enabled.
**test.tex
(./test.tex This is OpTeX (Olsak's Plain TeX), version <1.11 Mar 2023>
(./test.ref)
\openout1 = test.ref

(/usr/local/Cellar/texlive/20220321_4/share/texmf-dist/tex/optex/base/f-lmfonts
.opm
luaotfload | conf : Root cache directory is "/usr/local/Cellar/texlive/20220321_
4/share/texmf-var/luatex-cache/generic/names".
luaotfload | init : Loading fontloader "fontloader-2022-02-24.lua" from kpse-res
olved path "/usr/local/Cellar/texlive/20220321_4/share/texmf-dist/tex/luatex/lua
otfload/fontloader-2022-02-24.lua".
"luaotfload@noligature"=\attribute3
luaotfload | init : Context OpenType loader version 3.119
"luaotfload_color_attribute"=\attribute4
luaotfload | conf : Root cache directory is "/usr/local/Cellar/texlive/20220321_
4/share/texmf-var/luatex-cache/generic/names".
"luaotfload.letterspace_done"=\attribute5
luaotfload | db : Font names database loaded from /usr/local/Cellar/texlive/2022
0321_4/share/texmf-var/luatex-cache/generic/names/luaotfload-names.luc.gz
FONT: [Latin Modern] -- \LMfonts (TeX Gyre fonts based on Coputer Modern)
 mods:{\roman \sans \quotset \ttset \ttprop \ttlight \ttcond \upital \dunhill s
ubmods:\caps \slant \nbold \bolder } vars:{\rm \bf \it \bi \tt } math:{LM}
Modifiers (selects subfamily):
 \modreset .. implicit values (serif, bold extended, nocaps)
 \sans ...... sans serif subfamily
 \quotset ... quotation subfamily
 \ttset ..... typewriter subfamily
 \ttprop .... proportional typewriter-like subfamily
 \ttlight ... light typewriter subfamily
 \ttcond .... condensed typewriter subfamily
 \upital .... upright italic (only \rm )
 \dunhill ... dunhill roman subfamily (only \rm , \it )
Sub-modifiers (selects special variants of given subfamily):
 \caps ...... caps & small caps (available only at limited shapes)
 \slant ..... slanted (default in some subfamilies)
 \nbold ..... normal bold (extended bold is default)
 \bolder .... extended bold
Note: Text fonts support optical sizes, but LatinModern-Math only for scripts.
 Use another CM-like family if optical sizes not wanted: \fontfam [NewCM]. 
MATH-FONT: "[latinmodern-math]" -- unicode math prepared.

(/usr/local/Cellar/texlive/20220321_4/share/texmf-dist/tex/optex/base/unimath-c
odes.opm
@:[unimath-codes] Uni math codes <2023-01-17>
\_umathnumA=\count298
\_umathnumB=\count299

(/usr/local/Cellar/texlive/20220321_4/share/texmf-dist/tex/optex/base/mathclass.opm)
(/usr/local/Cellar/texlive/20220321_4/share/texmf-dist/tex/optex/base/unimath-table.opm)))
(/usr/local/Cellar/texlive/20220321_4/share/texmf-dist/tex/optex/base/usebib.opm
@:[usebib] Reading bib databases <2022-02-04>

(/usr/local/Cellar/texlive/20220321_4/share/texmf-dist/tex/generic/librarian/librarian.tex
\lb@entrycount=\count300
\lb@namecount=\count301
\lb@tempcount=\count302
\lb@sortnumber=\count303
\NameCount=\count304
\lb@read=\read2
))
(/usr/local/Cellar/texlive/20220321_4/share/texmf-dist/tex/optex/base/bib-simple.opm
**@:[bib-simple] BIB style (simple) <2022-10-10>
) (./biblist.bib)**
WARNING l.35: Missing field "journal" in [long].
WARNING l.35: Missing field "journal" in [short].
[1{/usr/local/Cellar/texlive/20220321_4/share/texmf-dist/fonts/map/pdftex/updmap
/pdftex.map}])</usr/local/Cellar/texlive/20220321_4/share/texmf-dist/fonts/opent
ype/public/lm/lmroman10-regular.otf>
```

I'm not sure why, though... I moved the files generated with `luatex -ini optex.ini` to where they should be, but maybe I was wrong. I'll look into it, but I'll mark this as solved anyway because your output is exactly what I was talking about. Thanks a lot!

**Reply by olsak** [reply]
> I'm not sure why, though... 

Perhaps you didn't copy new `bib-simple.opm` to `/usr/local/Cellar/texlive/20220321_4/share/texmf-dist/tex/optex/base/bib-simple.opm` (and analogically for `bib-iso690.opm`).`

**Reply by pabloabur** [reply]
I did, but with no luck. Is it possible that `bib-*.opm` from OpTeX [website](http://petr.olsak.net/ftp/olsak/optex/optex-111.zip) is not updated? I had a look at the source code and the preamble says

```
%% This is part of the OpTeX project, see http://petr.olsak.net/optex

\_codedecl \_undefined {BIB style (simple) <2022-10-10>} % loaded on demand by \usebib

\_ifx\_optexbibstyle\_undefined \_errmessage
   {This file can be read by: \_string\usebib/? (simple) bibfiles command only}
   \_endinput \_fi

% misc:
...
```

I am going to download the source code from GitHub to see if anything changes.

**Reply by pabloabur** [reply]
Yes, it works when I use github source code instead

**Reply by olsak** [reply]
I create a release sometimes (about per three months). You can see releases in the github commits history. On the other hand the unreleased actual version (with + at the end of the version number) is available only at github. The OpTeX web site says that the last zip file 111 (equivalent to the release 1.11) is from "9 Mar 2023" but my commit https://github.com/olsak/OpTeX/commit/2e5741bff4015d98ddec5891aa7f8a6e62e354c0 which solves your problem is from "24 March 2023". The new release 1.12 was not created.

**Reply by pabloabur** [reply]
I see. I totally missed that.

Just out of curiosity, in the code posted in optex tricks

```
   \def\_createbibmark #1;#2,#3;#4,#5;#6,#7;#8,#9\_fin{%
      \ifx ^#8^%
        \ifx ^#6^%
           \ifx ^#4^#2%
           \else #2, and #4\fi
        \else #2, #4, and #6\fi
      \else #2, et al.\fi
      \space (#1)%
   }
```

Does the argument surrounded by ^ (e.g. `^#8^`) check whether that argument is empty? I couldn't find this type of notation on my TeX references, so I don't know how this works

**Reply by olsak** [reply]
Yes, `\ifx ^#8^` is test, if `#8` is empty. When it is true then the code expands to `\ifx ^^` and the first token is equal to second token, the conditional returns true. If `#8` isn't empty and doesn't start with `^` (we are sure about this because the author name cannot begin with this math-mode operator), then the first token `^` isn't equal to the second token. The conditional returns false and skips following tokens to the relevant `\else` primitive.

**Reply by pabloabur** [reply]
Beautiful, I understand it now. Thanks!
